paymentsheet/src/main/java/com/stripe/android/paymentsheet/IntentConfirmationHandler.kt

```diff
@@ -0,0 +1,338 @@
+                    "No 'PaymentLauncher' instance was created before starting confirmation. Did you call register?"
+                ),
+                message = resolvableString(R.string.stripe_something_went_wrong),
+                type = ErrorType.Fatal,
+            )
+        }
+    }
+
+    private fun storeIsAwaitingForPaymentResult() {
+        savedStateHandle[AWAITING_PAYMENT_RESULT_KEY] = true
+    }
+
+    private fun removeIsAwaitingForPaymentResult() {
+        savedStateHandle.remove<Boolean>(AWAITING_PAYMENT_RESULT_KEY)
+    }
+
+    private fun isAwaitingForPaymentResult(): Boolean {
+        return savedStateHandle.get<Boolean>(AWAITING_PAYMENT_RESULT_KEY) ?: false
+    }
+
+    internal data class Args(
+        val initializationMode: PaymentSheet.InitializationMode,
+        val shippingDetails: AddressDetails?,
+    )
+
+    /**
+     * Defines the result types that [IntentConfirmationHandler] can return after completing the confirmation process.
+     */
+    sealed interface Result {
+        /**
+         * Indicates that the confirmation process was canceled by the customer.
+         */
+        data object Canceled : Result
+
+        /**
+         * Indicates that the confirmation process has been successfully completed. A [StripeIntent] with an updated
+         * state is returned as part of the result as well.
+         */
+        data class Succeeded(
+            val intent: StripeIntent,
+            val deferredIntentConfirmationType: DeferredIntentConfirmationType?
+        ) : Result
+
+        /**
+         * Indicates that the confirmation process has failed. A cause and potentially a resolvable message are
+         * returned as part of the result.
+         */
+        data class Failed(
+            val cause: Throwable,
+            val message: ResolvableString,
+            val type: ErrorType,
+        ) : Result
+    }
+
+    /**
+     * Types of errors that can occur when confirming an intent.
+     */
+    enum class ErrorType {
+        /**
+         * Fatal confirmation error that occurred while confirming a payment. This should never happen.
+         */
+        Fatal,
+
+        /**
+         * Indicates an error when processing a payment during the confirmation process.
+         */
+        Payment,
+
+        /**
+         * Indicates an error occurred when determining the next step for the confirmation process.
+         */
+        NextStep,
+    }
+
+    class Factory(
+        private val paymentSheetArguments: PaymentSheetContractV2.Args,
+        private val intentConfirmationInterceptor: IntentConfirmationInterceptor,
+        private val paymentConfigurationProvider: Provider<PaymentConfiguration>,
+        private val stripePaymentLauncherAssistedFactory: StripePaymentLauncherAssistedFactory,
+        private val savedStateHandle: SavedStateHandle,
+        private val application: Application,
+    ) {
+        fun create(scope: CoroutineScope): IntentConfirmationHandler {
```

Comment:
I'm wondering if we can get rid of the factory.

We can probably just create this inline in `PaymentSheetViewModel`. Not a blocker though. Probably leave it for now, and consider removing in a follow up.